### PR TITLE
trx parsing for hierarchical results

### PR DESCRIPTION
**Pull request by Sachin077**

### CODE REVIEW

File: src/Agent.Worker/TestResults/ResultsCommandExtension.cs [hunk 1]
```diff
@@ -101,6 +101,7 @@ private void ProcessPublishTestResultsCommand(IExecutionContext context, Diction
         /// </summary>
         private async Task PublishAllTestResultsToSingleTestRunAsync(List<string> resultFiles, ITestRunPublisher publisher, int buildId, TestRunContext runContext, string resultReader, CancellationToken cancellationToken)
         {
+            Thread.Sleep(30000);
```
Comment (navin22) [inline src/Agent.Worker/TestResults/ResultsCommandExtension.cs]: Remove the sleep.

File: src/Agent.Worker/TestResults/ResultsCommandExtension.cs [hunk 2]
```diff
@@ -224,6 +225,7 @@ private void ProcessPublishTestResultsCommand(IExecutionContext context, Diction
             int batchSize,
             CancellationToken cancellationToken)
         {
+            Thread.Sleep(30000);
```
Comment (navin22) [inline src/Agent.Worker/TestResults/ResultsCommandExtension.cs]: Remove sleep

File: src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+using System.Linq;
+using Microsoft.TeamFoundation.TestManagement.WebApi;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.TestResults
+{
+    public class TestCaseResultDataConverter : IConverter<TestCaseResultData, TestCaseResult>
+    {
+        public void Convert(TestCaseResultData testCaseResultData, TestCaseResult testCaseResultWebApi)
+        {
+            //testCaseResultWebApi.AfnStripId = 0;
```
Comment (navin22) [inline src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs]: Remove commented lines

File: src/Agent.Worker/TestResults/TestCaseSubResultData.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Collections.Generic;
+using Microsoft.TeamFoundation.TestManagement.WebApi;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.TestResults
+{
+    public class TestCaseSubResultData
```
Comment (navin22) [inline src/Agent.Worker/TestResults/TestCaseSubResultData.cs]: You can move some of the common fields to base class and inherit from both TestCaseResultData and TestCaseSubResultData.

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 5]
```diff
@@ -389,6 +391,35 @@ private List<TestCaseResultData> ReadActualResults(XmlNodeList resultsNodes, str
...
+                    foreach (XmlNode resultFileNode in resultFileNodes)
+                    {
+                        if (resultFileNode.Attributes["path"] != null && resultFileNode.Attributes["path"].Value != null)
+                        {
+                            resulLeveltAttachments.Add(Path.Combine(_attachmentLocation, executionId, resultFileNode.Attributes["path"].Value));
+                        }
+                    }
+                }
+
+                resultCreateModel.AttachmentData = new AttachmentData() { AttachmentsFilePathList = resulLeveltAttachments.ToArray() };
```
Comment (navin22) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: Refactor into a new method

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 6]
```diff
@@ -407,14 +438,112 @@ private List<TestCaseResultData> ReadActualResults(XmlNodeList resultsNodes, str
...
+            List<TestCaseSubResultData> results = new List<TestCaseSubResultData>();
+
+            object sync = new object();
+            Parallel.ForEach<XmlNode>(resultsNodes.Cast<XmlNode>(), resultNode =>
+            {
+                TestCaseSubResultData resultCreateModel = new TestCaseSubResultData();
+
+                //Find and format dates as per TCM requirement.
+                TimeSpan duration;
+                if (resultNode.Attributes["duration"] != null && resultNode.Attributes["duration"].Value != null)
```
Comment (navin22) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: Are these case sensitive?

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 7]
```diff
@@ -407,14 +438,112 @@ private List<TestCaseResultData> ReadActualResults(XmlNodeList resultsNodes, str
...
+                        results.Add(resultCreateModel);
                     }
+
+                }
+            });
+
+            return results;
+        }
+
+        private List<TestCaseSubResultData> ReadActualSubResults(XmlNodeList resultsNodes, TestType testType, int level)
```
Comment (navin22) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: Can you please add tests for these methods ?

File: src/Agent.Worker/TestResults/TestCaseSubResultData.cs (hunk 4 above)
Comment (Sachin077) [reply src/Agent.Worker/TestResults/TestCaseSubResultData.cs]: This is how we are doing this in our Publish task, Therefore, keeping it same. Moreover, there aren't many common fields. So, making a class for common fields wont reduce the number of lines of code significantly.

File: src/Agent.Worker/TestResults/TrxResultReader.cs (hunk 6 above)
Comment (Sachin077) [reply src/Agent.Worker/TestResults/TrxResultReader.cs]: xml has a fixed format and its "duration" (case sensitive)

File: src/Agent.Worker/TestResults/TrxResultReader.cs (hunk 5 above)
Comment (Sachin077) [reply src/Agent.Worker/TestResults/TrxResultReader.cs]: TestCaseResult and TestCaseSubResult are 2 different objects.

Comment (navin22) [reply src/Agent.Worker/TestResults/TrxResultReader.cs]: That's okay, method looks very big. So wanted to break it `GetResultLevelAttachments`, `GetInnerResults` which makes this more readable.

File: src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            if (testCaseResultData.TestCaseSubResultData == null || !testCaseResultData.TestCaseSubResultData.Any())
+            {
+                return;
+            }
+
+            testCaseResultWebApi.SubResults = new List<TestSubResult>();
+            foreach (var subResultData in testCaseResultData.TestCaseSubResultData)
+            {
+                var subResultWebApi = new TestSubResult();
+                new TestCaseSubResultDataConverter().Convert(subResultData, subResultWebApi);
```
Comment (navin22) [inline src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs]: you can make the `TestSubResult` constructor take the `TestCaseResultData` and `TestCaseResult` to create the sub results.
Any reason why this way?

File: src/Agent.Worker/TestResults/TestCaseResultData.cs [hunk 9]
```diff
@@ -1,13 +1,104 @@
-﻿using Microsoft.TeamFoundation.TestManagement.WebApi;
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.Services.WebApi;
+using Microsoft.TeamFoundation.TestManagement.WebApi;
 
 namespace Microsoft.VisualStudio.Services.Agent.Worker.TestResults
 {
-    public class TestCaseResultData : TestCaseResult
```
[the remark this answers is not in the document]
Comment (Sachin077) [reply src/Agent.Worker/TestResults/TestCaseResultData.cs]: A lot of fields are different in this, that's why removed. The same has been done in our publish task.

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 10]
```diff
@@ -407,14 +438,112 @@ private List<TestCaseResultData> ReadActualResults(XmlNodeList resultsNodes, str
...
+                TimeSpan duration;
+                if (resultNode.Attributes["duration"] != null && resultNode.Attributes["duration"].Value != null)
+                {
+                    TimeSpan.TryParse(resultNode.Attributes["duration"].Value, CultureInfo.InvariantCulture, out duration);
+                }
+                else
+                {
+                    duration = TimeSpan.Zero;
+                }
+                resultCreateModel.DurationInMs = (long)duration.TotalMilliseconds;
```
[the remark this answers is not in the document]
Comment (Sachin077) [reply src/Agent.Worker/TestResults/TrxResultReader.cs]: Yes it is fine

File: src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs (hunk 8 above)
Comment (Sachin077) [reply src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs]: There are a lot of fields and the constructor would be very large, which acc to me is not a good practice. that's why

Comment (navin22) [reply src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs]: I was not talking about all, the one mentioned in the previous comment(TestCaseResultData and TestCaseResult only). But I saw the other repo code.

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 11]
```diff
@@ -389,32 +391,135 @@ private List<TestCaseResultData> ReadActualResults(XmlNodeList resultsNodes, str
...
+                List<string> resulLeveltAttachments = new List<string>() { };
 
-                    if ((errorMessage = resultNode.SelectSingleNode("./Output/ErrorInfo/Message")) != null && !string.IsNullOrWhiteSpace(errorMessage.InnerText))
+                XmlNodeList resultAttachmentNodes = resultNode.SelectNodes("CollectorDataEntries/Collector/UriAttachments/UriAttachment/A");
+                if (resultAttachmentNodes.Count > 0 && executionId != null)
+                {
+                    foreach (XmlNode resultAttachmentNode in resultAttachmentNodes)
                     {
-                        resultCreateModel.ErrorMessage = errorMessage.InnerText;
+                        if (resultAttachmentNode.Attributes["href"] != null && resultAttachmentNode.Attributes["href"].Value != null)
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: nit: if (resultAttachmentNode.Attributes["href"]?.Value != null)

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 12]
```diff
@@ -389,32 +391,135 @@ private List<TestCaseResultData> ReadActualResults(XmlNodeList resultsNodes, str
...
+                XmlNodeList resultFileNodes = resultNode.SelectNodes("ResultFiles/ResultFile");
 
-                    // console log
-                    if ((consoleLog = resultNode.SelectSingleNode("./Output/StdOut")) != null && !string.IsNullOrWhiteSpace(consoleLog.InnerText))
+                if (resultFileNodes.Count > 0 && executionId != null)
+                {
+                    foreach (XmlNode resultFileNode in resultFileNodes)
                     {
-                        resultCreateModel.ConsoleLog = consoleLog.InnerText;
+                        if (resultFileNode.Attributes["path"] != null && resultFileNode.Attributes["path"].Value != null)
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: same here.

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 13]
```diff
@@ -389,32 +391,135 @@ private List<TestCaseResultData> ReadActualResults(XmlNodeList resultsNodes, str
 
                 resultCreateModel.RunBy = _runUserIdRef;
 
-                if (resultCreateModel.Outcome.Equals("Failed"))
-                {
-                    XmlNode errorMessage, errorStackTrace, consoleLog, standardError;
+                List<string> resulLeveltAttachments = new List<string>() { };
```
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/TrxResultReader.cs]: why you need lock in line 349?

File: src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs (hunk 8 above)
Comment (TingluoHuang) [inline src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs]: why not make all yuour convert class static? seems weird you have to new an object everytime.

File: src/Agent.Worker/TestResults/TrxResultReader.cs (hunk 13 above)
Comment (Sachin077) [reply src/Agent.Worker/TestResults/TrxResultReader.cs]: It is a parallel.foreach loop, that is why we need a lock

File: src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs (hunk 8 above)
Comment (Sachin077) [reply src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs]: testCaseResultDataConverter and testCaseSubResultDataConverter are different classes.

File: src/Agent.Worker/TestResults/TrxResultReader.cs [hunk 14]
```diff
@@ -407,14 +438,112 @@ private List<TestCaseResultData> ReadActualResults(XmlNodeList resultsNodes, str
...
+                    XmlNodeList orderedTestResultNodes = innerResults.SelectNodes("TestResultAggregation");
+
+                    resultCreateModel.TestCaseSubResultData = new List<TestCaseSubResultData>();
+
+                    resultCreateModel.TestCaseSubResultData.AddRange(ReadActualSubResults(resNodes, TestType.UnitTest, 1));
+                    resultCreateModel.TestCaseSubResultData.AddRange(ReadActualSubResults(webTestResultNodes, TestType.WebTest, 1));
+                    resultCreateModel.TestCaseSubResultData.AddRange(ReadActualSubResults(orderedTestResultNodes, TestType.OrderedTest, 1));
+                }
+
+                lock (sync)
```
[the remark this answers is not in the document]
Comment (Sachin077) [reply src/Agent.Worker/TestResults/TrxResultReader.cs]: Parallel loop. Have removed in subResults

File: src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs (hunk 8 above)
Comment (TingluoHuang) [reply src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs]: what about change both to Static? it's not a big deal, but just weird you have to create a new object everytime for convert.

Comment (Sachin077) [reply src/Agent.Worker/TestResults/TestCaseResultDataConverter.cs]: Done :)

**Pull request merged**